utils/datasets/cifar10_dataset.py

```diff
@@ -0,0 +1,348 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library for loading and preprocessing CIFAR-10 training and testing data."""
+
+import collections
+from typing import Tuple
+import numpy as np
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+TEST_SAMPLES_PER_CLIENT = 1000
+NUM_CLASSES = 10
+NUM_CLIENTS = 10
+
+def load_cifar10_federated(dirichlet_parameter: float = 1):
+  '''Construct a synthetic federated CIFAR-10 from its centralized version
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+  Args:
+    dirichlet_parameter: A float parametrizing the Dirichlet distribution which samples a 
+      multinomial over classes for each client.
+  Returns:
+    A tuple of `tff.simulation.ClientData` representing unpreprocessed
+    train data and test data.
+  '''
+  (train_images, train_labels), (test_images, test_labels) \
+      = datasets.cifar10.load_data()
+  train_clients = collections.OrderedDict()
+  test_clients = collections.OrderedDict()
+
+  
+  train_multinomial_vals = []
+  test_multinomial_vals = []
+  # each column is a distribution over classes for the client at that index
+  for i in range(num_clients):
+      proportion = np.random.dirichlet(dirichlet_parameter*np.ones(NUM_CLASSES,))
+      train_multinomial_vals.append(proportion)
+      test_multinomial_vals.append(proportion)
+        
+  train_multinomial_vals = np.array(train_multinomial_vals)
+  test_multinomial_vals = np.array(test_multinomial_vals)
+  
+  train_indices = []
+  test_indices = []
+  for k in range(NUM_CLASSES):
+    train_label_k = np.where(train_labels==k)[0]
+    np.random.shuffle(train_label_k)
+    train_indices.append(train_label_k)
+    test_label_k = np.where(test_labels==k)[0]
+    np.random.shuffle(test_label_k)
+    test_indices.append(test_label_k)
+
+  train_indices = np.array(train_indices)
+  test_indices = np.array(test_indices)
+
+  train_client_samples = [[] for _ in range(NUM_CLIENTS)]
+  test_client_samples = [[] for _ in range(NUM_CLIENTS)]
+  '''
+  sample 5000 images for the current client according 
+  to the multinomial without replacement
+  '''
+  train_count = np.zeros(NUM_CLASSES).astype(int)
+  test_count = np.zeros(NUM_CLASSES).astype(int)
+
+  for k in range(NUM_CLASSES):
```

Comment:
should this be NUM_CLIENTS?

Reply:
I think it should be NUM_CLASSES and we would have to change NUM_EXAMPLES_PER_CLIENT if we want to change the number of clients.